File: coffee-vending-machine-java/src/main/java/barista/CoffeeFactory.java
package barista;

import barista.coffee.Espresso;
import barista.coffee.HouseBlend;

public class CoffeeFactory {
    private CoffeeFactory() {
    }

    private static volatile CoffeeFactory factory;

    public static CoffeeFactory getInstance() {
        if (factory == null) {
            synchronized (CoffeeFactory.class) {
                if (factory == null) {
                    factory = new CoffeeFactory();
                }
            }
        }
        return factory;
    }

    public Beverage serverCoffee(CoffeeType type, Boolean hasMilk, Boolean hasSyrup, Boolean hasSugar) {
        switch (type){
            case ESPRESSO: {
                Espresso.Builder builder = new Espresso.Builder();

                if (hasMilk) builder.milk();
                if (hasSyrup) builder.syrup();
                if (hasSugar) builder.sugar();

                return builder.build();
            }
            case HOUSE_BLEND: {
                HouseBlend.Builder builder = new HouseBlend.Builder();

                if (hasMilk) builder.milk();
                if (hasSyrup) builder.syrup();
                if (hasSugar) builder.sugar();

                return builder.build();
            }
            // Note: no compile time checking of branches
            default: {
                throw new IllegalArgumentException("Incorrect choice provided!");
            }
        }
    }
}

File: coffee-vending-machine-kotlin/src/main/kotlin/barista/CoffeeFactory.kt
package barista

import barista.coffee.Espresso
import barista.coffee.HouseBlend
import barista.condiments.FlavoredSyrup
import barista.condiments.Milk
import barista.condiments.Sugar

object CoffeeFactory {
    fun getCoffee(coffeeType: CoffeeType, hasMilk: Boolean?, hasSyrup: Boolean?, hasSugar: Boolean?, quantity: Int = 1): List<Beverage> {
        // Note: Compile time checking of all code branches
        return when (coffeeType) {
            CoffeeType.HOUSE_BLEND -> (1..quantity).map {
                HouseBlend(milk = hasMilk?.let { Milk() },
                        syrup = hasSyrup?.let { FlavoredSyrup() },
                        sugar = hasSugar?.let { Sugar() })
            }
            CoffeeType.ESPRESSO -> (1..quantity).map {
                Espresso(milk = hasMilk?.let { Milk() },
                        syrup = hasSyrup?.let { FlavoredSyrup() },
                        sugar = hasSugar?.let { Sugar() })
            }
        }
    }
}
